DEEP_DIVES/package-managers/README.md

```diff
@@ -0,0 +1,121 @@
+  20. [Not as many central points of failure — or of trust](decentralizedpub-fewercentralpoints.jpg)
+
+
+## Key takeaways from ranking exercise
+
+For the dot-ranking exercise, we discussed each of the 20 pain points/potential benefits individually, but focused the discussion on the items in which the dots were divergent between participants. Since our participants came from a variety of backgrounds and areas of focus -- particularly in that we had a mix of package manager users of various backgrounds and preferred languages, as well as a package manager maintainer -- the most interesting nuances were to be found in the areas where opinions differed. See the below for key takeaways.
+*Source: @achingbrain*
+
+- Connectivity is important, eg. working offline
+- Replication of content not important to end users
+- New paradigms of package management enable new use cases
+- Reproducibility! Very important
+- Decentralised publishing is a great goal
+- Decentralised registries not massively important
+- Excessive network activity is slow, user hosted packages would help
+- Why refetch assets/tarballs/etc that we know will not have changed? - e.g. redownloading is bad
+- Performance is not that important as long as it's 'fast enough'
+- Performance regressions the worst
+- Left-pad type issues are not common but are show stoppers when they do
+- Being able to reproduce builds when dependencies change over time very important
+- Cost of hosting private/internal registries not important
+- Deduping package contents important for developers but less so for maintainers, particularly if they dedupe already
+- Package creators care more about self-hosting than package users
+- All participants liked the option to fork, but not very often -- it's most useful if need to modify software for yourself
+
+## Guix patch notes that may be helpful in other package manager efforts
+*Source: @momack2*
+
+- Want executable bit
+- Create an IPLD node with that metadata using CBOR with a pointer to package and whatever key values 
+- Flexible and can get going right now 
+- Add stuff normally to IPFS, then reference in JSON file with other metadata, then hash that and get a new root CID
+- Easier because working in schema
+- Add to unixfs v1 using metadata 
+- If you make the file executable, is it a different file? Using unixfs metadata -- they'd be different
+- Tar the result, send to IPFS, then chunk it
+- Would need to improve chunker in go-ipfs
+- Need to make the tar chunker reproducible
+- When rebuilding the tarball out of IPFS, might not reconstruct the tarball metadata perfectly
+- Take the gradual approach: This is smooth from user perspective, because can look at multiple registries
+- Can do on a per-package basis
+- Pin just the root instead of each at once
+
+## Next steps
+- Discuss the dot-rankings and how they might impact overall prioritization efforts on the package manager roadmap
```

Comment:
```suggestion
- Discuss the dot-rankings and how they might impact overall prioritization efforts on the package manager roadmap

## UnixFS v1.5?

Hands down our most requested feature from package manager maintainers is to add persistent metadata to UnixFS files.  There is an unused [metadata](https://github.com/ipfs/js-ipfs-unixfs/blob/master/src/unixfs.proto.js#L22-L24) construct in the spec but implementing it would have the side-effect of changing the file's CID when it's metadata changes.

Building on @hsanjuan's suggestion we could enable the majority of use cases by creating a dag-CBOR node that contains unstructured metadata and a link to a UnixFS file.  Call it UnixFSv1.5.

The `content` linked to could be a UnixFSv1 file/dir/hamt/whatever which also gives users an upgrade path without having to re-encode all of their data:

```javascript
{
  mtime: 1561879007581
  perm: '0777',
  // ... more metadata here
  content: {
    '/': 'Qmfoo' // link to a UnixFSv1 file or dir or hamt or cbor, etc..
  }
}
```